File: product-brain/src/product_brain/blocks/edge_mine.py
from __future__ import annotations

import json
import re
import subprocess
from collections import Counter, defaultdict
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from ..models import Commit, EdgeCaseBullet, PullRequest, TicketRecord


_VERB_PATTERNS = re.compile(
    r"\b(fix(?:es|ed)?|handle[sd]?|guard|edge case|race|deadlock|leak|null|nil|undefined|panic|crash|retry|fallback|timeout|race condition)\b",
    re.IGNORECASE,
)
_TODO_RE = re.compile(r"(?://|#|/\*)\s*TODO[\s\(:]")
_TEST_DEF = re.compile(r"^\s*(?:def|fn|func|it|test)[\s(]+(test_[a-zA-Z0-9_]+|[a-zA-Z0-9_]*Test[a-zA-Z0-9_]*)")


@dataclass
class Signals:
    pr_review_comments: list[dict]
    added_test_names: list[dict]
    commit_verb_lines: list[dict]
    added_code_comments: list[dict]
    pm_description: str
    linked_bug_tickets: list[str]


def gather_signals(
    repo_path: Path,
    commits: list[Commit],
    prs: list[PullRequest],
    pm_description: str,
    linked_bugs: Optional[list[str]] = None,
) -> Signals:
    pr_comments = []
    for pr in prs:
        for c in pr.review_comments:
            pr_comments.append({
                "author": c.author, "body": c.body, "file": c.file, "line": c.line, "pr": pr.number,
            })

    test_names: list[dict] = []
    for c in commits:
        for f in c.files:
            if f.change != "added":
                continue
            if not _looks_like_test(f.path):
                continue
            for line in _read_blob_at(repo_path, c.sha, f.path).splitlines():
                m = _TEST_DEF.match(line)
                if m:
                    test_names.append({"name": m.group(1), "file": f.path, "sha": c.sha})

    verb_lines = []
    for c in commits:
        text = c.subject + "\n" + (c.body or "")
        for line in text.splitlines():
            if _VERB_PATTERNS.search(line):
                verb_lines.append({"line": line.strip(), "sha": c.sha})

    code_comments = _scan_diff_comments(repo_path, commits)

    return Signals(
        pr_review_comments=pr_comments,
        added_test_names=test_names,
        commit_verb_lines=verb_lines,
        added_code_comments=code_comments,
        pm_description=pm_description or "",
        linked_bug_tickets=linked_bugs or [],
    )


def _looks_like_test(path: str) -> bool:
    p = path.lower()
    return ("test_" in p or "_test." in p or p.startswith("test/") or "/test/" in p
            or "/tests/" in p or ".test." in p or ".spec." in p)


def _read_blob_at(repo: Path, sha: str, path: str) -> str:
    try:
        r = subprocess.run(
            ["git", "-C", str(repo), "show", f"{sha}:{path}"],
            capture_output=True, text=True, check=True, errors="replace",
        )
        return r.stdout
    except subprocess.CalledProcessError:
        return ""


def _scan_diff_comments(repo: Path, commits: list[Commit]) -> list[dict]:
    out = []
    for c in commits:
        try:
            r = subprocess.run(
                ["git", "-C", str(repo), "show", "--unified=0", "--no-color", c.sha],
                capture_output=True, text=True, check=True, errors="replace",
            )
        except subprocess.CalledProcessError:
            continue
        cur_file = None
        for line in r.stdout.splitlines():
            if line.startswith("+++"):
                cur_file = line[6:].strip() if line.startswith("+++ b/") else line[4:].strip()
            elif line.startswith("+") and not line.startswith("+++"):
                t = line[1:].lstrip()
                if t.startswith(("//", "#", "/*")) and "TODO" not in t.upper() and len(t) > 8:
                    out.append({"text": t[:240], "file": cur_file or "", "sha": c.sha})
                elif _TODO_RE.search(t):
                    out.append({"text": t[:240], "file": cur_file or "", "sha": c.sha, "kind": "todo"})
    return out


_EXTRACT_PROMPT = """You extract edge cases that were considered or handled in this ticket.

Use ONLY the SIGNALS below. For each edge case:
  - Quote or closely paraphrase from a signal.
  - Cite the source EXACTLY as one of: "pr#N review @user", "test_name", "commit <sha7>", or "<path>:<line> TODO".
  - If signals don't support N bullets, return fewer (or zero).
  - DO NOT extrapolate from the feature description.

Output STRICT JSON, no commentary:
{
  "what_shipped": "one paragraph from PR/commits, no invention",
  "key_decisions": ["..."],
  "edge_cases_handled": [{"text": "...", "source": "..."}],
  "known_gaps": [{"text": "...", "source": "..."}]
}

SIGNALS:
%(signals)s
"""


def extract_with_llm(signals: Signals, llm_call) -> dict:
    payload = {
        "pr_review_comments": signals.pr_review_comments[:50],
        "added_test_names": signals.added_test_names[:50],
        "commit_verb_lines": signals.commit_verb_lines[:50],
        "added_code_comments": signals.added_code_comments[:50],
        "pm_description": signals.pm_description[:2000],
        "linked_bug_tickets": signals.linked_bug_tickets,
    }
    prompt = _EXTRACT_PROMPT % {"signals": json.dumps(payload, indent=2)}
    raw = llm_call(prompt)
    try:
        return json.loads(_first_json_object(raw))
    except (ValueError, json.JSONDecodeError):
        return {"what_shipped": "", "key_decisions": [], "edge_cases_handled": [], "known_gaps": []}


def _first_json_object(s: str) -> str:
    start = s.find("{")
    if start == -1:
        return "{}"
    depth = 0
    for i in range(start, len(s)):
        if s[i] == "{":
            depth += 1
        elif s[i] == "}":
            depth -= 1
            if depth == 0:
                return s[start:i + 1]
    return s[start:]


def validate_citations(
    bullets: list[EdgeCaseBullet],
    repo_path: Path,
    pr_numbers_seen: set[int],
) -> tuple[list[EdgeCaseBullet], int]:
    kept: list[EdgeCaseBullet] = []
    dropped = 0
    for b in bullets:
        if _validate_one(b.source, repo_path, pr_numbers_seen):
            kept.append(b)
        else:
            dropped += 1
    return kept, dropped


def _validate_one(source: str, repo: Path, pr_numbers: set[int]) -> bool:
    s = source.strip()
    m = re.match(r"pr#(\d+)", s)
    if m:
        return int(m.group(1)) in pr_numbers
    m = re.match(r"commit ([0-9a-f]{7,40})", s)
    if m:
        try:
            subprocess.run(["git", "-C", str(repo), "cat-file", "-e", m.group(1)],
                           check=True, capture_output=True)
            return True
        except subprocess.CalledProcessError:
            return False
    if s.startswith("test_") or "::" in s:
        name = s.split("::")[-1]
        try:
            r = subprocess.run(
                ["git", "-C", str(repo), "grep", "-l", name],
                capture_output=True, text=True, check=False,
            )
            return bool(r.stdout.strip())
        except FileNotFoundError:
            return False
    m = re.match(r"(.+?):(\d+)\s+TODO", s)
    if m:
        path = repo / m.group(1)
        return path.exists()
    return False


def mine_per_ticket(
    repo_path: Path,
    commits: list[Commit],
    prs: list[PullRequest],
    pm_description: str,
    llm_call,
    linked_bugs: Optional[list[str]] = None,
) -> tuple[dict, int]:
    signals = gather_signals(repo_path, commits, prs, pm_description, linked_bugs)
    extracted = extract_with_llm(signals, llm_call)

    pr_numbers = {pr.number for pr in prs}
    raw_edges = [EdgeCaseBullet(text=e["text"], source=e["source"])
                 for e in extracted.get("edge_cases_handled", []) if e.get("source")]
    raw_gaps = [EdgeCaseBullet(text=e["text"], source=e["source"])
                for e in extracted.get("known_gaps", []) if e.get("source")]
    edges, dropped_e = validate_citations(raw_edges, repo_path, pr_numbers)
    gaps, dropped_g = validate_citations(raw_gaps, repo_path, pr_numbers)

    return {
        "what_shipped": extracted.get("what_shipped", ""),
        "key_decisions": extracted.get("key_decisions", []),
        "edge_cases_handled": edges,
        "known_gaps": gaps,
    }, (dropped_e + dropped_g)


def dedup_edge_cases(records: list[TicketRecord], llm_call=None) -> list[dict]:
    """Cluster bullets across related records.

    Without LLM: groups by token-set Jaccard similarity.
    With LLM: optional pass to produce themed labels.
    """
    bullets: list[tuple[EdgeCaseBullet, str]] = []
    for r in records:
        for b in r.edge_cases_handled:
            bullets.append((b, r.ticket))

    groups: list[list[tuple[EdgeCaseBullet, str]]] = []
    for entry in bullets:
        b, _ = entry
        toks = _normalize_tokens(b.text)
        placed = False
        for grp in groups:
            ref_toks = _normalize_tokens(grp[0][0].text)
            if _token_jaccard(toks, ref_toks) >= 0.45:
                grp.append(entry)
                placed = True
                break
        if not placed:
            groups.append([entry])

    out = []
    for grp in groups:
        tickets = sorted({tid for _, tid in grp})
        rep = grp[0][0]
        out.append({
            "text": rep.text,
            "frequency": f"{len(tickets)}/{len(records)} records",
            "tickets": tickets,
            "sources": [b.source for b, _ in grp],
        })
    out.sort(key=lambda d: -len(d["tickets"]))
    return out


def _normalize_tokens(s: str) -> set[str]:
    return {t.lower() for t in re.findall(r"[a-zA-Z]{4,}", s)}


def _token_jaccard(a: set, b: set) -> float:
    if not a and not b:
        return 0.0
    return len(a & b) / max(len(a | b), 1)

File: product-brain-ts/src/blocks/edge-mine.ts
// Port target: ../product-brain/src/product_brain/blocks/edge_mine.py
import { execSync } from "node:child_process";
import { existsSync } from "node:fs";
import path from "node:path";
import type {
  Commit,
  EdgeCaseBullet,
  PullRequest,
  RunResult,
  TestCase,
  TicketRecord,
} from "../models.js";

const VERB_RE =
  /\b(fix(?:es|ed)?|handle[sd]?|guard|edge case|race|deadlock|leak|null|nil|undefined|panic|crash|retry|fallback|timeout|race condition)\b/i;
const TODO_RE = /(?:\/\/|#|\/\*)\s*TODO[\s(:]/;
const TEST_DEF_RE =
  /^\s*(?:def|fn|func|it|test)[\s(]+(test_[a-zA-Z0-9_]+|[a-zA-Z0-9_]*Test[a-zA-Z0-9_]*)/;

export interface Signals {
  prReviewComments: Array<Record<string, unknown>>;
  addedTestNames: Array<Record<string, unknown>>;
  commitVerbLines: Array<Record<string, unknown>>;
  addedCodeComments: Array<Record<string, unknown>>;
  pmDescription: string;
  linkedBugTickets: string[];
  testCases: Array<Record<string, unknown>>;
  testRunHistory: Array<Record<string, unknown>>;
}

function looksLikeTest(p: string): boolean {
  const lower = p.toLowerCase();
  return (
    lower.includes("test_") ||
    lower.includes("_test.") ||
    lower.startsWith("test/") ||
    lower.includes("/test/") ||
    lower.includes("/tests/") ||
    lower.includes(".test.") ||
    lower.includes(".spec.")
  );
}

function readBlobAt(repo: string, sha: string, p: string): string {
  try {
    return execSync(`git -C "${repo}" show "${sha}:${p}"`, { encoding: "utf8" });
  } catch {
    return "";
  }
}

function scanDiffComments(repo: string, commits: Commit[]): Array<Record<string, unknown>> {
  const out: Array<Record<string, unknown>> = [];
  for (const c of commits) {
    let stdout: string;
    try {
      stdout = execSync(`git -C "${repo}" show --unified=0 --no-color "${c.sha}"`, {
        encoding: "utf8",
        maxBuffer: 32 * 1024 * 1024,
      });
    } catch {
      continue;
    }
    let curFile = "";
    for (const line of stdout.split("\n")) {
      if (line.startsWith("+++")) {
        curFile = line.startsWith("+++ b/") ? line.slice(6).trim() : line.slice(4).trim();
      } else if (line.startsWith("+") && !line.startsWith("+++")) {
        const t = line.slice(1).replace(/^\s+/, "");
        if (
          (t.startsWith("//") || t.startsWith("#") || t.startsWith("/*")) &&
          !t.toUpperCase().includes("TODO") &&
          t.length > 8
        ) {
          out.push({ text: t.slice(0, 240), file: curFile, sha: c.sha });
        } else if (TODO_RE.test(t)) {
          out.push({ text: t.slice(0, 240), file: curFile, sha: c.sha, kind: "todo" });
        }
      }
    }
  }
  return out;
}

export function gatherSignals(
  repoPath: string,
  commits: Commit[],
  prs: PullRequest[],
  pmDescription: string,
  opts: {
    linkedBugs?: string[];
    testCases?: TestCase[];
    runHistory?: RunResult[];
  } = {},
): Signals {
  const prReviewComments: Array<Record<string, unknown>> = [];
  for (const pr of prs) {
    for (const c of pr.reviewComments) {
      prReviewComments.push({
        author: c.author,
        body: c.body,
        file: c.file,
        line: c.line,
        pr: pr.number,
      });
    }
  }

  const addedTestNames: Array<Record<string, unknown>> = [];
  for (const c of commits) {
    for (const f of c.files) {
      if (f.change !== "added" || !looksLikeTest(f.path)) continue;
      for (const line of readBlobAt(repoPath, c.sha, f.path).split("\n")) {
        const m = TEST_DEF_RE.exec(line);
        if (m) addedTestNames.push({ name: m[1], file: f.path, sha: c.sha });
      }
    }
  }

  const commitVerbLines: Array<Record<string, unknown>> = [];
  for (const c of commits) {
    const text = `${c.subject}\n${c.body ?? ""}`;
    for (const line of text.split("\n")) {
      if (VERB_RE.test(line)) commitVerbLines.push({ line: line.trim(), sha: c.sha });
    }
  }

  return {
    prReviewComments,
    addedTestNames,
    commitVerbLines,
    addedCodeComments: scanDiffComments(repoPath, commits),
    pmDescription: pmDescription || "",
    linkedBugTickets: opts.linkedBugs ?? [],
    testCases: (opts.testCases ?? []).map((c) => ({
      id: c.id,
      title: c.title,
      automation: c.automation,
      type: c.type,
      last_status: c.lastStatus,
      recent_failures: c.recentFailures,
    })),
    testRunHistory: (opts.runHistory ?? []).map((r) => ({
      case_id: r.caseId,
      status: r.status,
      timestamp: r.timestamp ? r.timestamp.toISOString() : null,
      comment: r.comment,
    })),
  };
}

const EXTRACT_PROMPT = `Extract edge cases handled in this ticket.

Rules:
- Use ONLY SIGNALS below. No invention.
- Quote or close-paraphrase from a signal.
- Cite source EXACTLY: "pr#N review @user" | "test_name" | "commit <sha7>" | "<path>:<line> TODO" | "TR-C-NNNN".
- QA edges (from test_cases): case title must be specific. Skip generic ("test login works").
- Signals don't support N bullets → return fewer / zero. Do not pad.
- Do not extrapolate from feature description.

Output STRICT JSON only:
{
  "what_shipped": "one paragraph from PR/commits",
  "key_decisions": ["..."],
  "edge_cases_handled": [{"text": "...", "source": "..."}],
  "known_gaps": [{"text": "...", "source": "..."}],
  "qa_edges": [{"text": "...", "source": "TR-C-NNNN (automation, last_status)"}]
}

SIGNALS:
%SIGNALS%
`;

function firstJsonObject(s: string): string {
  const start = s.indexOf("{");
  if (start === -1) return "{}";
  let depth = 0;
  for (let i = start; i < s.length; i++) {
    if (s[i] === "{") depth += 1;
    else if (s[i] === "}") {
      depth -= 1;
      if (depth === 0) return s.slice(start, i + 1);
    }
  }
  return s.slice(start);
}

export type LLMCall = (prompt: string, maxTokens?: number) => Promise<string>;

export async function extractWithLLM(signals: Signals, llmCall: LLMCall): Promise<Record<string, unknown>> {
  const payload = {
    pr_review_comments: signals.prReviewComments.slice(0, 50),
    added_test_names: signals.addedTestNames.slice(0, 50),
    commit_verb_lines: signals.commitVerbLines.slice(0, 50),
    added_code_comments: signals.addedCodeComments.slice(0, 50),
    pm_description: signals.pmDescription.slice(0, 2000),
    linked_bug_tickets: signals.linkedBugTickets,
    test_cases: signals.testCases.slice(0, 50),
    test_run_history: signals.testRunHistory.slice(0, 50),
  };
  const prompt = EXTRACT_PROMPT.replace("%SIGNALS%", JSON.stringify(payload, null, 2));
  const raw = await llmCall(prompt);
  try {
    return JSON.parse(firstJsonObject(raw)) as Record<string, unknown>;
  } catch {
    return {
      what_shipped: "",
      key_decisions: [],
      edge_cases_handled: [],
      known_gaps: [],
      qa_edges: [],
    };
  }
}

function validateOne(
  source: string,
  repo: string,
  prNumbers: Set<number>,
  testCaseIds: Set<string>,
): boolean {
  const s = source.trim();
  let m = /^pr#(\d+)/.exec(s);
  if (m) return prNumbers.has(Number(m[1]));
  m = /^commit ([0-9a-f]{7,40})/.exec(s);
  if (m) {
    try {
      execSync(`git -C "${repo}" cat-file -e ${m[1]}`, { stdio: "ignore" });
      return true;
    } catch {
      return false;
    }
  }
  m = /^(TR-C-\d+)/.exec(s);
  if (m) return testCaseIds.has(m[1] ?? "");
  if (s.startsWith("test_") || s.includes("::")) {
    const name = s.split("::").pop() ?? s;
    try {
      const out = execSync(`git -C "${repo}" grep -l "${name}"`, { encoding: "utf8" });
      return out.trim().length > 0;
    } catch {
      return false;
    }
  }
  m = /^(.+?):(\d+)\s+TODO/.exec(s);
  if (m) return existsSync(path.join(repo, m[1] ?? ""));
  return false;
}

export function validateCitations(
  bullets: EdgeCaseBullet[],
  repoPath: string,
  prNumbersSeen: Set<number>,
  knownTestCaseIds: Set<string> = new Set(),
): { kept: EdgeCaseBullet[]; dropped: number } {
  const kept: EdgeCaseBullet[] = [];
  let dropped = 0;
  for (const b of bullets) {
    if (validateOne(b.source, repoPath, prNumbersSeen, knownTestCaseIds)) kept.push(b);
    else dropped += 1;
  }
  return { kept, dropped };
}

export function stabilitySignals(testCases: TestCase[], runs: RunResult[]): string[] {
  const out: string[] = [];
  const byCase = new Map<string, RunResult[]>();
  for (const r of runs) {
    const arr = byCase.get(r.caseId) ?? [];
    arr.push(r);
    byCase.set(r.caseId, arr);
  }
  const titles = new Map(testCases.map((c) => [c.id, c.title]));
  for (const [caseId, results] of byCase) {
    const failures = results.filter((r) => r.status === "failed" || r.status === "blocked");
    if (failures.length >= 3) {
      const title = titles.get(caseId) ?? "(no linked case in this record)";
      out.push(
        `${caseId} (${title}): ${failures.length} failures/blocks in window — structurally fragile`,
      );
    }
  }
  for (const c of testCases) {
    if (c.recentFailures >= 3 && !out.some((s) => s.includes(c.id))) {
      out.push(`${c.id} (${c.title}): ${c.recentFailures} recent failures`);
    }
  }
  return out;
}

interface MineResult {
  what_shipped: string;
  key_decisions: string[];
  edge_cases_handled: EdgeCaseBullet[];
  known_gaps: EdgeCaseBullet[];
  qa_edges: EdgeCaseBullet[];
  stability_signals: string[];
}

export async function minePerTicket(
  repoPath: string,
  commits: Commit[],
  prs: PullRequest[],
  pmDescription: string,
  llmCall: LLMCall,
  opts: {
    linkedBugs?: string[];
    testCases?: TestCase[];
    runHistory?: RunResult[];
  } = {},
): Promise<{ result: MineResult; dropped: number }> {
  const signals = gatherSignals(repoPath, commits, prs, pmDescription, opts);
  const extracted = await extractWithLLM(signals, llmCall);

  const prNumbers = new Set(prs.map((p) => p.number));
  const caseIds = new Set((opts.testCases ?? []).map((c) => c.id));

  const toBullets = (key: string): EdgeCaseBullet[] => {
    const arr = (extracted[key] as Array<Record<string, unknown>>) ?? [];
    return arr
      .filter((e) => e["source"])
      .map((e) => ({ text: String(e["text"] ?? ""), source: String(e["source"]) }));
  };

  const e = validateCitations(toBullets("edge_cases_handled"), repoPath, prNumbers, caseIds);
  const g = validateCitations(toBullets("known_gaps"), repoPath, prNumbers, caseIds);
  const q = validateCitations(toBullets("qa_edges"), repoPath, prNumbers, caseIds);

  return {
    result: {
      what_shipped: String(extracted["what_shipped"] ?? ""),
      key_decisions: ((extracted["key_decisions"] as unknown[]) ?? []).map(String),
      edge_cases_handled: e.kept,
      known_gaps: g.kept,
      qa_edges: q.kept,
      stability_signals: stabilitySignals(opts.testCases ?? [], opts.runHistory ?? []),
    },
    dropped: e.dropped + g.dropped + q.dropped,
  };
}

function normalizeTokens(s: string): Set<string> {
  return new Set((s.match(/[a-zA-Z]{4,}/g) ?? []).map((t) => t.toLowerCase()));
}

function tokenJaccard(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 && b.size === 0) return 0;
  let inter = 0;
  for (const x of a) if (b.has(x)) inter += 1;
  const union = a.size + b.size - inter;
  return inter / Math.max(union, 1);
}

export interface DedupGroup {
  text: string;
  frequency: string;
  tickets: string[];
  sources: string[];
}

export function dedupEdgeCases(records: TicketRecord[]): DedupGroup[] {
  const bullets: Array<[EdgeCaseBullet, string]> = [];
  for (const r of records) {
    for (const b of r.edgeCasesHandled) bullets.push([b, r.ticket]);
  }
  const groups: Array<Array<[EdgeCaseBullet, string]>> = [];
  for (const entry of bullets) {
    const toks = normalizeTokens(entry[0].text);
    let placed = false;
    for (const grp of groups) {
      const refToks = normalizeTokens(grp[0]?.[0].text ?? "");
      if (tokenJaccard(toks, refToks) >= 0.45) {
        grp.push(entry);
        placed = true;
        break;
      }
    }
    if (!placed) groups.push([entry]);
  }
  const out: DedupGroup[] = groups.map((grp) => {
    const tickets = [...new Set(grp.map(([, tid]) => tid))].sort();
    const first = grp[0];
    return {
      text: first?.[0].text ?? "",
      frequency: `${tickets.length}/${records.length} records`,
      tickets,
      sources: grp.map(([b]) => b.source),
    };
  });
  out.sort((a, b) => b.tickets.length - a.tickets.length);
  return out;
}
